### PR TITLE
Add boolean selection to _select_columns

### PR DESCRIPTION
**Pull request by samukweku**
# PR Description

Please describe the changes proposed in the pull request:

- Allow selection of list of booleans.
- Ensures only one data type(list) is returned, to avoid confusion.
- improve select_columns/pivot_longer/pivot_wider to accept list-likes.
- add a `level` parameter to `_select_columns`, primarily for `pivot_longer` and MultiIndexes.
- add a check in `select_columns`, limiting it to only single index columns.

**This PR improve  #790.**

# PR Checklist

Please ensure that you have done the following:

1. [x] PR in from a fork off your branch. Do not PR from `<your_username>`:`dev`, but rather from `<your_username>`:`<feature-branch_name>`.

2. [x] If you're not on the contributors list, add yourself to `AUTHORS.rst`.

3. [x] Add a line to `CHANGELOG.md` under the latest version header (i.e. the one that is "on deck") describing the contribution.
    - Do use some discretion here; if there are multiple PRs that are related, keep them in a single line.

## Quick Check

To do a very quick check that everything is correct, follow these steps below:

- [x] Run the command `make check` from pyjanitor's top-level directory. This will automatically run:
    - black formatting
    - flake8 checking
    - running the test suite
    - docs build

Once done, please check off the check-box above.

If `make check` does not work for you, you can execute the commands listed in the Makefile individually.

## Code Changes

If you are adding code changes, please ensure the following:

- [x] Ensure that you have added tests.
- [x] Run all tests (`$ pytest .`) locally on your machine.
    - [x] Check to ensure that test coverage covers the lines of code that you have added.
    - [x] Ensure that all tests pass.

## Documentation Changes

If you are adding documentation changes, please ensure the following:

- [x] Build the docs locally.
- [x] View the docs to check that it renders correctly.

# Relevant Reviewers

Please tag maintainers to review.

- @ericmjl

### CODE REVIEW

**Comment by ericmjl** [comment]
Ahhh! I see, this PR supersedes #790, is that right?

**Comment by samukweku** [comment]
@ericmjl Yes it does

**Comment by ericmjl** [comment]
Got it, @samukweku! Looks like only a Jupyter notebook is failing at the moment. I will leave a review shortly too.

File: janitor/functions.py [hunk 1]
```diff
@@ -3227,6 +3240,11 @@ def select_columns(
         [str, callable, Pattern, slice, list],
     )
 
+    # applicable for any
+    # list-like object (ndarray, Series, pd.Index, tuple, ...)
+    if pd.api.types.is_list_like(search_column_names):
```
Comment (ericmjl) [inline janitor/functions.py]: Oooh, cool catch! I didn't know the pandas Types API provided this haha.

File: tests/utils/test__select_column.py [hunk 2]
```diff
@@ -148,6 +162,13 @@ def test_regex_presence(df):
         _select_columns(re.compile(r"^\d+"), df)
 
 
+@pytest.mark.xfail(
+    reason="""
+            no need to test for tuples,
+            as any function using `_select_columns`
+            should convert any list-like to lists.
+            """
```
Comment (ericmjl) [inline tests/utils/test__select_column.py]: Awesome! Thanks for marking it with a reason, @samukweku :smile:.

File: tests/utils/test__select_column.py [hunk 3]
```diff
@@ -170,96 +191,88 @@ def test_strings(df1):
...
-        df1.loc[:, slice(None, "code2")].columns,
-    )
-    assert_index_equal(_select_columns(slice(None, None), df1), df1.columns)
-    assert_index_equal(
-        _select_columns(slice(None, None, 2), df1),
-        df1.loc[:, slice(None, None, 2)].columns,
-    )
+    assert _select_columns(slice("code", "code2"), df1), df1.loc[
+        :, slice("code", "code2")
+    ].columns.tolist()
```
Comment (ericmjl) [inline tests/utils/test__select_column.py]: @samukweku, I think I'm having a bit of trouble following this test, can you help me a bit here? Are you trying to assert equality between the result of `_select_columns` (in the first item position) and the second item? If so, I think you'll need to use the `assert pos1 == pos2` pattern, unless I'm missing something with the assert statement.

**Review by ericmjl — APPROVED** [review]
@samukweku, just one question regarding the assertions in the test suite; other than that, I'm pre-approving for merging later (don't want to be the bottleneck here!).

File: janitor/functions.py (hunk 1 above)
Comment (samukweku) [reply janitor/functions.py]: @eric, pandas has some very rich features that makes life easy :); yeah, not really sure; I wouldnt want to tamper too much with someone else's code; I think I'll take this off, and if users are interested in it, maybe include it.

Comment (ericmjl) [reply janitor/functions.py]: Actually, let's not take it off, I think it's a great idea :smile:.

File: tests/utils/test__select_column.py (hunk 3 above)
Comment (samukweku) [reply tests/utils/test__select_column.py]: realised assert does not need the `==` (it worked :) ), but I have fixed it in the accepted format.

**Comment by ericmjl** [comment]
@samukweku are we ready to merge?

**Comment by samukweku** [comment]
@ericmjl I will do One more check tonight and then merge

**Comment by samukweku** [comment]
@ericmjl , all good on my end

**Pull request merged**